aiokafka/record/legacy_records.py

```diff
@@ -278,155 +285,155 @@ def __repr__(self):
+            return LegacyRecordMetadata(offset, crc, size, timestamp)
+
+        except struct.error:
+            # perform expensive type checking only to translate struct errors
+            # to human-readable messages
+            if type(offset) != int:
+                raise TypeError(offset)
+            if type(timestamp) != int:
+                raise TypeError(timestamp)
+            if not isinstance(key, (bytes, bytearray, memoryview, NoneType)):
+                raise TypeError("Unsupported type for key: %s" % type(key))
+            if not isinstance(value, (bytes, bytearray, memoryview, NoneType)):
+                raise TypeError("Unsupported type for value: %s" % type(value))
+            raise
+
+    def _encode_msg(self, start_pos, size, offset, timestamp, key_size, key,
+                    value_size, value, attributes=0):
         """ Encode msg data into the `msg_buffer`, which should be allocated
             to at least the size of this message.
         """
         magic = self._magic
-        buf = self._buffer
-        pos = start_pos
+        buf = memoryview(self._buffer)[start_pos:]
 
-        # Write key and value
-        pos += self.KEY_OFFSET_V0 if magic == 0 else self.KEY_OFFSET_V1
+        length = (self.KEY_LENGTH + key_size
+                  + self.VALUE_LENGTH + value_size
+                  - self.LOG_OVERHEAD)
 
-        if key is None:
-            struct.pack_into(">i", buf, pos, -1)
-            pos += self.KEY_LENGTH
-        else:
-            key_size = len(key)
-            struct.pack_into(">i", buf, pos, key_size)
-            pos += self.KEY_LENGTH
-            buf[pos: pos + key_size] = key
-            pos += key_size
-
-        if value is None:
-            struct.pack_into(">i", buf, pos, -1)
-            pos += self.VALUE_LENGTH
-        else:
-            value_size = len(value)
-            struct.pack_into(">i", buf, pos, value_size)
-            pos += self.VALUE_LENGTH
-            buf[pos: pos + value_size] = value
-            pos += value_size
-        length = (pos - start_pos) - self.LOG_OVERHEAD
-
-        # Write msg header. Note, that Crc will be updated later
         if magic == 0:
-            self.HEADER_STRUCT_V0.pack_into(
-                buf, start_pos,
-                offset, length, 0, magic, attributes)
+            length += self.KEY_OFFSET_V0
+            struct.pack_into(
+                ">q"   # BaseOffset => Int64
+                "i"    # Length => Int32
+                "I"    # CRC => Int32
+                "b"    # Magic => Int8
+                "b"    # Attributes => Int8
+                "i"    # key length => Int32
+                "%ds"  # key => bytes
+                "i"    # value length => Int32
+                "%ds"  # value => bytes
+                % (key_size, value_size),
+                buf, 0, offset, length, 0, magic, attributes,
+                key_size or -1, key or b"", value_size or -1, value or b"")
```

Comment:
There's an error here. The `key_size or -1` will give a false result on `key == b""`. We had to write empty bytes but resulted in null bytes.

Reply:
Ahh, great catch. Fixed and added a test for this case.